test/impl_copied.cpp

```diff
@@ -34,7 +34,7 @@ template<> struct boost::impl_ptr<Copied>::implementation
     mutable string trace_;
 };
 
-Copied::Copied ()      : impl_ptr_type(in_place) {}
+Copied::Copied ()      : impl_ptr_type(std::allocator_arg, std::allocator<void>()) {}
```

Comment:
> The simple default constructor seems replaced with a considerably more complex one. Are you proposing that we force the user to always provide allocator-specific arguments (as in the line)?

No, I just hijacked this one example line in this test as an example for using an allocator. The default, with the `in_place` tag should still be preferred by users. I'll change this and just add another test case focused exclusively on allocators instead to avoid this confusion.

> To have "b)" the allocator needs to be passed as a parameter and the allocator to be stored with implementation (as it is done in boost::shared_ptr, see sp_counted_impl_pda)... and that is how you've (I believe implemented it). If so, then I'd expect allocator-related constructors to be in addition to the existing ones. However, from the code, it appears, you replace instead. Am I wrong?

I assume you're referring to the removal of the constructor overloads taking `detail::in_place_type` from the policy types. Yes, I am replacing those with constructors taking allocators by default. But if you look at `impl_ptr`, which is the user's interface to these policies, you'll find that I provide wrappers for:

1. neither `in_place` nor an allocator
2. `in_place`
3. an allocator

So the internals become simpler (all in terms of the allocator), while the external interface maintains its current flexibility. Then the allocator-related constructors and emplacement functions get added to the existing ones, not replacing anything. (Though the implementations of the existing ones do change, referring to the allocator ones).

Reply:
Thanks for the explanations. Much appreciated. I merged your changes (I believe you could do it yourself). Looking at the changes closer now... A hell of a job!.. Allocators are probably one of the darkest, poorly documented corners. Very impressive.